signac/contrib/import_export.py

```diff
@@ -241,6 +241,14 @@ def path_function(job):
             except Exception as error:
                 raise _SchemaPathEvaluationError(error)
 
+        # Check that the user-specified path generates a 1-1 mapping
+        link_check = {j.workspace(): path_function(j) for j in jobs}
+        if len(set(link_check.values())) != len(link_check):
+            raise RuntimeError(
+                "Links generated with the given path function are not 1-1. The easiest "
+                "way to fix this is to add the job id to the path specification."
+            )
```

Comment:
I suggest the following changes:
- Raise the error for the _first_ failure case rather than evaluating the paths for all jobs before erroring out
- Show the path causing a failure in the error message.
- Use a `ValueError` since it's the user-provided path value (optional, feel free to keep `RuntimeError`)
```suggestion
        links = set()
        for job in jobs:
            job_path = path_function(job)
            if job_path in links:
                raise ValueError(
                    f"Multiple jobs resolve to the same path '{job_path}'. "
                    "The easiest way to fix this is to add the job id to the "
                    "path specification."
                )
            links.add(job_path)
```

Reply:
+1 for using a set, it doesn't look like the dict key matters here. I prefer constructing the entire set at once though (via something like `links = set(path_function(job) for job in jobs)`) rather than erroring on the first duplicate. The main reason is that the error message in the suggestion is potentially misleading: it indicates that there is only a single conflict, when in fact there may (most likely there _will_) be multiple. Even worse, if you have a heterogeneous schema the paths may be underspecified for different reasons. If we really want to provide the most useful diagnostic (but it's a bit more work and I'd understand if you wanted to skip), we should use the for loop, use a flag to indicate that duplicates were found, then log _all_ duplicates using the logger so that the user can inspect the problem. Barring that, I think that constructing the set in one shot and not showing a specific path is probably more appropriate here.

As a bonus, the single-pass approach will be faster in the normal case (no failure). The generator-based construction will be faster than a for loop, you won't need to do a containment check on every add, and the final test is a simple length check

Reply:
Thanks @bdice @vyasr !
I combined both approaches. Thanks for the advice to use `set()`.

I like being able to see colliding paths, but not by default, in case there are 1000s of them. I put them in the debug log.

I am wondering if it would be faster for most cases as @vyasr points out to first do the simple length check and then only iterate through all jobs if that fails? This would result in duplicated code but could be faster for huge workspaces?

Reply:
@vyasr I'm a little unsure about what preference you're expressing in your statement: 
> Barring that, I think that constructing the set in one shot and not showing a specific path is probably more appropriate here.

Now that the "most useful diagnostic" is implemented with a for loop, would you prefer that to the simpler error that does not show a specific path? I think you make good points here and it aligns with what is done for exports: https://github.com/glotzerlab/signac/blob/60646fdf29e5aa28680a6462337ab5a0488bf6c6/signac/contrib/import_export.py#L328-L333

Reply:
@bdice sorry for the confusion, let me clarify. My ordered prefs were:
1. Check if lengths mismatch, report all unmatched paths in the debug log via `logger.debug(...)` and then raise an exception
2. Check if lengths mismatch and raise an error without specifying paths
3. Loop through and find all mismatches, then raise an error with all of them if any exist

As far as I can tell, what is currently implemented is option (2), which is the simpler case that you found also in `import_export.py`. @cbkerr I'm confused by your last message:

> I like being able to see colliding paths, but not by default, in case there are 1000s of them. I put them in the debug log.
What did you put in the debug log? I don't see any logging right now.

> I am wondering if it would be faster for most cases as @vyasr points out to first do the simple length check and then only iterate through all jobs if that fails? This would result in duplicated code but could be faster for huge workspaces?
The length check is already what you're doing, right? You just aren't doing the iteration at all?

Reply:
Agree that what's released is option 2. The code I added is option 1 except that it does not do the simple length check first.

> I do think that the best option is to start with a length check first.

I think you're agreeing with what I was asking about here:
> to first do the simple length check and then only iterate through all jobs if that fails

Reply:
Yes, I am agreeing with that.